### PR TITLE
Set some advice/standards for test linting

### PR DESCRIPTION
**Pull request by JGAntunes**
What I am suggesting:

- [x] Gets out of the way;
- [x] Is not restricting;
- [x] Is safe, production ready, and project members have experience with it;
- [x] Promotes good coding practices.

^ Only used these because they're on the template but I'm not entirely sure if this applies to this specific PR. Anyway the point is giving some advice on using airbnb linting rules when linting tests, as it won't run straight out of the box for the tools we're used to (mocha, chai, etc.).

### CODE REVIEW

**Comment by nfcampos** [comment]
i'd strongly suggest staying away from mocha as a test runner but i guess enough projects use it that it's useful to have rules specifically for it?

**Comment by JGAntunes** [comment]
This is not about mocha but more about a specific set of rules you'll
probably need to lint your tests. Mocha was just an example.

**Comment by nfcampos** [comment]
yes but mocha is the only test runner that uses this for setting timeout...

**Comment by nfcampos** [comment]
i guess what i mean is that rules that apply only for certain dependencies should be more explicitly worded to reflect that, eg. something like "If you use chai then disable no-unused-expressions"

File: js-code-style.md [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
 We can start a project to extend one of those bases if we find we have issues with some rules.
 
+### For tests
+
+You should fine-tune the config for your tests folder.
+First set the right environment in your eslint config (`mocha` for example).
+There's also a set of rules that you should probably ignore/change. Consider the following:
+
+  * `no-unused-expressions`: Should be `off` when using assertion libs like `chai` for example.
+  * `prefer-arrow-callback`: Should be `off` when using test frameworks like mocha as stated [in their documentation](https://mochajs.org/#arrow-functions).
```
Comment (tomgco) [inline js-code-style.md:14]: I disagree with this, having consistency between our code and tests is paramount.

File: js-code-style.md [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
+  * `func-names`: Taking into account the above, it is just painful to name all of your test functions
```
Comment (tomgco) [inline js-code-style.md:15]: See Above ^^

File: js-code-style.md [hunk 3]
```diff
@@ -1,9 +1,20 @@
 # Javascript Code Style
 
-## eslint with the `eslint-config-airbnb-base` preset, `eslint-config-airbnb` for React projects;
+## ESLint with the `eslint-config-airbnb-base` preset, `eslint-config-airbnb` for React projects;
 
 We can start a project to extend one of those bases if we find we have issues with some rules.
 
+### For tests
```
Comment (tomgco) [inline js-code-style.md:7]: Maybe something about setting the correct `NODE_ENV` as an explicit step?

Comment (tomgco) [reply js-code-style.md:7]: I really think that having a standard test output format (TAP) is much better than a custom reporter. This allows anyone to format test output as the developer wishes.

**Review by tomgco — CHANGES_REQUESTED** [review]
A few things that I personally disagree with here. (Changing lint specifically for tests)

**Comment by JGAntunes** [comment]
I spoke with @nfcampos yesterday and I agree that maybe we should either drop this or probably rephrase it to something more of a -  "Rules/configuration steps to watch out when using mocha/jest/lab/tape/ava/<fancy-test-runner>".

File: js-code-style.md (hunk 3 above)
Comment (JGAntunes) [reply js-code-style.md:7]: Agree with all of it. But should this be in the JS Code Style?

File: js-code-style.md (hunk 1 above)
Comment (JGAntunes) [reply js-code-style.md:14]: For me is just a matter of people that use test runners such as mocha (not sure if anymore suffer from the same problems) to be aware of the set backs of using arrow functions. If everyone is one board with it I'm okay too :+1:
Also, the context of this whole discussion should probably change I think - see https://github.com/yldio/standards/pull/5#issuecomment-262253587

Comment (JGAntunes) [reply js-code-style.md:14]: Although I tend to disagree with this specific rule in some scenarios, for which I think it might fail, but that will be a problem for another discussion :smile:

Comment (joaojeronimo) [reply js-code-style.md:14]: If this illustrates that mocha is not a good choice of a test runner even better, I hope we can set a standard test runner that everyone agrees will play well with linters and other tools.

**Comment by JGAntunes** [comment]
@joaojeronimo I feel like that's too much of an Utopian dream :smile:

João Antunes

File: js-code-style.md (hunk 1 above)
Comment (JGAntunes) [reply js-code-style.md:14]: @joaojeronimo I feel like that's too much of an Utopian dream :smile:

**Comment by joaojeronimo** [comment]
So what's the status of this ?

**Comment by JGAntunes** [comment]
I'll rephrase some of the stuff here, to change from "mandatory" to advisory when using stuff like mocha or chai, etc. is that ok? Or I can also close this, since it may not be considered much of a standard but more of a set of advice on how to deal with the airbnb eslint config. - @nfcampos @joaojeronimo

**Comment by joaojeronimo** [comment]
Choose one and

**Review by tomgco — APPROVED** [review]

**Comment by tomgco** [comment]
Didn't realise the review was blocking merges. I am happy, sorry 💃 🗡
